### PR TITLE
NEW - See the documents of my subordinates 

### PR DESCRIPTION
**Pull request by BB2A-Anthony**
# NEW|New [*See the documents of my subordinates *]
[*This function is essential in my opinion*]

proposal following comment from PR #24035

@eldy if good for you i change for propal invoice and thirdparty in other PR

### CODE REVIEW

**Comment by eldy** [comment]
This will generate a lot of trouble in permission systems and data view will not be consistent between screen, but this is not a pb as long as it  remains a hidden option.
However what is your definition of "order of subordinates" ?
- Order that are on thirdparties with one contact that is an external users that is among subordinates ?
- Or order that has a sale representative that is among subordinates ?

**Comment by BB2A-Anthony** [comment]
> This will generate a lot of trouble in permission systems and data view will not be consistent between screen, but this is not a pb as long as it remains a hidden option. However what is your definition of "order of subordinates" ?
> 
>     * Order that are on thirdparties with one contact that is an external users that is among subordinates ?
> 
>     * Or order that has a sale representative that is among subordinates ?

for me it's no longer this case "order that has a sale representative that is among subordinates"

What do you mean by 
"This will generate a lot of trouble in permission systems and data view will not be consistent between screen"
Can you give me more details?

**Comment by eldy** [comment]
> > This will generate a lot of trouble in permission systems and data view will not be consistent between screen, but this is not a pb as long as it remains a hidden option. However what is your definition of "order of subordinates" ?
> > ```
> > * Order that are on thirdparties with one contact that is an external users that is among subordinates ?
> > 
> > * Or order that has a sale representative that is among subordinates ?
> > ```
> 
> for me it's no longer this case "order that has a sale representative that is among subordinates"
> 
> What do you mean by "This will generate a lot of trouble in permission systems and data view will not be consistent between screen" Can you give me more details?

100% of the code is base in this rule:
- Permission on HR modules are based on perm + subordinates
- Permission on project/tasks are based on perm + who is assigned
- Permission on all the rest is based on perm only
Here you introduce a rule that is a mix of A and C, so this will generate troubles. User will surely not understand why sometimes he can see and sometimes he can not see some data.

If your definition is "order that has a sale representative that is among subordinates", the modification must not be done by changing the sql at each place by by forcing the search_salerepresentative to the list of id of sale representative that are subordinates.

So the code that you have removed
```
if (!$user->hasRight('societe', 'client', 'voir')) {
	$search_sale = $user->id;
}
```
should not be removed, but replaced with 
```
if (!$user->hasRight('societe', 'client', 'voir')) {
	$search_sale = list of allowed id,...,...,...
}
```
So we force the filter at begin of code in the head/security section.

then in sql instead of doing a filter on search_sale, it must be a filter on IN ($db->sanitize($search_sale))

**Comment by BB2A-Anthony** [comment]
> $search_sale

I understand. for me it doesn't make sense to mix research with security.

$search_sale lives up to its name and is made for search.

If I display the orders of my subordinates, how can I filter them to one of them or myself. Since the field is preselected with my name.

The goal is to have this behavior on all views and the user will not be lost since he knows that he is the supervisor. It is normal for a supervisor to see the work of his subordinates.

I agree that this should be an option to keep the initial behavior by default.

**Comment by BB2A-Anthony** [comment]
> > > This will generate a lot of trouble in permission systems and data view will not be consistent between screen, but this is not a pb as long as it remains a hidden option. However what is your definition of "order of subordinates" ?
> > > ```
> > > * Order that are on thirdparties with one contact that is an external users that is among subordinates ?
> > > 
> > > * Or order that has a sale representative that is among subordinates ?
> > > ```
> > 
> > 
> > for me it's no longer this case "order that has a sale representative that is among subordinates"
> > What do you mean by "This will generate a lot of trouble in permission systems and data view will not be consistent between screen" Can you give me more details?
> 
> 100% of the code is base in this rule:
> 
>     * Permission on HR modules are based on perm + subordinates
> 
>     * Permission on project/tasks are based on perm + who is assigned
> 
>     * Permission on all the rest is based on perm only
>       Here you introduce a rule that is a mix of A and C, so this will generate troubles. User will surely not understand why sometimes he can see and sometimes he can not see some data.
> 
> 
> If your definition is "order that has a sale representative that is among subordinates", the modification must not be done by changing the sql at each place by by forcing the search_salerepresentative to the list of id of sale representative that are subordinates.
> 
> So the code that you have removed
> 
> ```
> if (!$user->hasRight('societe', 'client', 'voir')) {
> 	$search_sale = $user->id;
> }
> ```
> 
> should not be removed, but replaced with
> 
> ```
> if (!$user->hasRight('societe', 'client', 'voir')) {
> 	$search_sale = list of allowed id,...,...,...
> }
> ```
> 
> So we force the filter at begin of code in the head/security section.
> 
> then in sql instead of doing a filter on search_sale, it must be a filter on IN ($db->sanitize($search_sale))


I understand. for me it doesn't make sense to mix research with security.

$search_sale lives up to its name and is made for search.

If I display the orders of my subordinates, how can I filter them to one of them or myself. Since the field is preselected with my name.

The goal is to have this behavior on all views and the user will not be lost since he knows that he is the supervisor. It is normal for a supervisor to see the work of his subordinates.

I agree that this should be an option to keep the initial behavior by default.

There is also a security problem because if you clear the filters it displays the complete list without restriction.

It is necessary that
```
if (!$user->hasRight('societe', 'client', 'voir')) {
	$search_sale = $user->id;
}
```

Or lower.

**Comment by BB2A-Anthony** [comment]
@eldy I rewrote the code to be simpler and optimized on the principles of holidays like in your example.

File: htdocs/commande/list.php [hunk 1]
```diff
@@ -879,6 +883,16 @@
 if ($socid > 0) {
 	$sql .= ' AND s.rowid = '.((int) $socid);
 }
+
+// Restriction on sale representative
+if (!$permissiontoreadallthirdparty && !$socid) {
```
[the remark this answers is not in the document]
Comment (eldy) [reply htdocs/commande/list.php:888]: Also the condition on MAIN_SEE_SUBORDINATES must appears.

[the remark this answers is not in the document]
Comment (BB2A-Anthony) [reply htdocs/commande/list.php:888]: Ok it's done

File: htdocs/commande/list.php [hunk 2]
```diff
@@ -223,15 +227,16 @@
 //$arrayfields['anotherfield'] = array('type'=>'integer', 'label'=>'AnotherField', 'checked'=>1, 'enabled'=>1, 'position'=>90, 'csslist'=>'right');
 $arrayfields = dol_sort_array($arrayfields, 'position');
 
-if (!$user->hasRight('societe', 'client', 'voir')) {
-	$search_sale = $user->id;
```
Comment (eldy) [inline htdocs/commande/list.php:227]: It seems we loose this protection when we are in default mode:
If MAIN_SEE_SUBORDINATES not set and use has permission to read all, tis was used to restrict list on $search_sale at line 1011 (in old code file)
So you should restore a restriction into the select. For example in your addition line 888
`if (!$permissiontoreadallthirdparty && getDolGlobalInt('MAIN_SEE_SUBORDINATES')) {`
you can do
```
if (!$permissiontoreadallthirdparty) {
 if (getDolGlobalInt('MAIN_SEE_SUBORDINATES')) {
      // restrict on list in userschilds
} else {
   // default mode, restrict on user->id  only.


```

Comment (BB2A-Anthony) [reply htdocs/commande/list.php:227]: yes I see I did

**Pull request merged**